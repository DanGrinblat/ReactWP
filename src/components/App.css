.pagination {
	margin-bottom: 78px !important;
}

 
