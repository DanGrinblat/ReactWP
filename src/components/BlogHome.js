import React from "react";
import PostList from "./PostList";
import Pagination from "./Pagination";
import wprest from "../api/wprest";

class BlogHome extends React.Component {
	state = {
		posts: [],
		media: [],
		categories: [],
		pageCount: 1,
		postCount: 0,
		postsPerPage: 6,
		selectedPage: 1
	};

	componentDidMount() {
		this.retrieveAll(this.state.selectedPage, this.state.postsPerPage);
	}

	onClickPage = newPage => {
		this.setState({
			selectedPage: Number(newPage)
		});
		this.retrieveAll(newPage, this.state.postsPerPage);
	};

	//Retrieve Categories, Posts, and Media in succession
	retrieveAll = async (page, perPage) => {
		const categoriesData = await this.retrieveCategories();
		const postsData = await this.retrievePosts(
			page,
			perPage,
			categoriesData
		);

		const postIdsList = postsData.map(post => {
			return post.featured_media;
		});
		await this.retrieveMedia(postIdsList);
	};

	retrievePosts = async (page, perPage, categories) => {
		var params = { page: page, per_page: perPage };

		const selectedCategory = this.props.match.params.categoryterm;
		if (selectedCategory != null) {
			var selectedCategoryExists = false;
			for (var i = 0; i < categories.length; i++) {
				if (selectedCategory === categories[i].slug) {
					params.categories = [categories[i].id];
					selectedCategoryExists = true;
					break;
				}
			}
			if (!selectedCategoryExists) {
				return [];
			}
		}

		const response = await wprest.get("wp/v2/posts", { params });
		console.log(response);

		if (this.state.posts !== response.data) {
			//If there are 0 pages, set pageCount to 1
			const newPageCount =
				response.headers["x-wp-totalpages"] !== "0"
					? response.headers["x-wp-totalpages"]
					: "1";
			this.setState({
				posts: response.data,
				postCount: response.headers["x-wp-total"],
				pageCount: newPageCount
			});
		}
		return response.data;
	};

	async retrieveMedia(postIdsList) {
		const response = await wprest.get("wp/v2/media", {
			params: { include: postIdsList }
		});

		if (this.state.media !== response.data) {
			this.setState({ media: response.data });
		}
		return response.data;
	}

	async retrieveCategories() {
		const response = await wprest.get("wp/v2/categories");
		if (this.state.categories !== response.data) {
			this.setState({ categories: response.data });
		}
		return response.data;
	}

	render() {
		return (
			<div>
				<PostList posts={this.state.posts} media={this.state.media} />
				<center>
					<Pagination
						selectedPage={this.state.selectedPage}
						pageCount={this.state.pageCount}
						postCount={this.state.postCount}
						onClickPage={this.onClickPage}
					/>
				</center>
			</div>
		);
	}
}

export default BlogHome;
