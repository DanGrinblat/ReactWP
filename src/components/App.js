import "./App.css";
import React from "react";
import Header from "./Header";
import Footer from "./Footer";
import BlogHome from "./BlogHome";
import SinglePost from "./SinglePost";
import { Router, Route } from "react-router-dom";
import history from "./history";

class App extends React.Component {
	render() {
		return (
			<div style={{ height: "100vh" }}>
				<Header />
				<Router history={history}>
					<Route path="/" exact component={BlogHome} />
					<Route
						path="/category/:categoryterm"
						exact
						component={BlogHome}
					/>
					<Route
						path="/blog/:postname"
						exact
						component={SinglePost}
					/>
				</Router>
				<Footer />
			</div>
		);
	}
}

export default App;

//Home - Should Display Latest # Posts ( Title and excerpt only ) with pagination
