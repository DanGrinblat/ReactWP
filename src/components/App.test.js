import React from "react";
import { shallow } from "enzyme";
import { Route } from "react-router";
import BlogHome from "./BlogHome";
import App from "./App";

describe("App", () => {
	const app = shallow(<App />);

	it("should render correctly", () => {
		expect(app).toMatchSnapshot();
	});

	describe("when creating Routes", () => {
		let pathMap;

		beforeEach(() => {
			pathMap = app.find(Route).reduce((pathMap, route) => {
				const routeProps = route.props();
				pathMap[routeProps.path] = routeProps;
				return pathMap;
			}, {});
		});

		it("should create a Route to BlogHome", () => {
			expect(pathMap["/"].component).toBe(BlogHome);
		});

		it("should create a Route to BlogHome with URL param `test` set as key", () => {
			const props = { match: { params: { categoryterm: "test" } } };

			const blogHomeWrapper = shallow(
				pathMap["/category/:categoryterm"].render(props)
			);

			expect(
				blogHomeWrapper.instance().props.match.params.categoryterm
			).toEqual("test");
		});
	});
});
