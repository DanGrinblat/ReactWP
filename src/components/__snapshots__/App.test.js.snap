// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<div
  class="app-full-screen"
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Header />
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      exact={true}
      path="/category/:categoryterm"
      render={[Function]}
    />
    <Route
      component={[Function]}
      exact={true}
      path="/blog/:postslug"
    />
    <Footer />
  </Router>
</div>
`;
